From 2da31a067fb342490631f41042ff3ef8a6d716ef Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Wed, 19 Nov 2025 15:51:11 +0100
Subject: [PATCH] i2c: stm32f7: ignore spurious bus-error when acting as
 controller

When acting as controller, spurious bus error can be generated.
Even if the BERR raised, the transfer is not impacted and only handling
necessary is to clear the flag.

Change-Id: I6ffd141bf5b2fde26694c17caea0c62d59c4ebb7
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/510690
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
---
 drivers/i2c/busses/i2c-stm32f7.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/drivers/i2c/busses/i2c-stm32f7.c b/drivers/i2c/busses/i2c-stm32f7.c
index 5d84eb85d085..20e8045baaf3 100644
--- a/drivers/i2c/busses/i2c-stm32f7.c
+++ b/drivers/i2c/busses/i2c-stm32f7.c
@@ -1525,10 +1525,16 @@ static irqreturn_t stm32f7_i2c_handle_isr_errs(struct stm32f7_i2c_dev *i2c_dev,
 
 	/* Bus error */
 	if (status & STM32F7_I2C_ISR_BERR) {
-		dev_err(dev, "Bus error accessing addr 0x%x\n", addr);
 		writel_relaxed(STM32F7_I2C_ICR_BERRCF, base + STM32F7_I2C_ICR);
-		stm32f7_i2c_release_bus(&i2c_dev->adap);
-		f7_msg->result = -EIO;
+		/*
+		 * Spurious bus error can be received when running as controller.
+		 * In such situation, transfer is not impacted and can proceed.
+		 */
+		if (!i2c_dev->master_mode) {
+			dev_err(dev, "Bus error accessing addr 0x%x\n", addr);
+			stm32f7_i2c_release_bus(&i2c_dev->adap);
+			f7_msg->result = -EIO;
+		}
 	}
 
 	/* Arbitration loss */
@@ -1548,9 +1554,12 @@ static irqreturn_t stm32f7_i2c_handle_isr_errs(struct stm32f7_i2c_dev *i2c_dev,
 		dev_dbg(dev, "SMBus alert received\n");
 		writel_relaxed(STM32F7_I2C_ICR_ALERTCF, base + STM32F7_I2C_ICR);
 		i2c_handle_smbus_alert(i2c_dev->alert->ara);
-		return IRQ_HANDLED;
 	}
 
+	/* Not all flags leads to need to terminate the transfer */
+	if (!f7_msg->result)
+		return IRQ_HANDLED;
+
 	if (!i2c_dev->slave_running) {
 		u32 mask;
 		/* Disable interrupts */
-- 
2.53.0

